### ISSUE TITLE
New user friendly authentication

### ISSUE THREAD

**Issue by jolestar** [rooch-network/rooch#769]
- [ ] https://joy.id/
- [ ] https://sui.io/zklogin

**Comment by yubing744** [comment]
I'll try this task

**Comment by yubing744** [comment]
@jolestar I drew a data flow diagram using Rooch to implement zkLogin (privacy is not guaranteed):

![image](https://github.com/rooch-network/rooch/assets/1449069/cf221aed-2af1-4b94-98d5-318a5cd61523)

**Comment by yubing744** [comment]
Rooch implements the OAuth login process:

1. The user authenticates their identity through the OAuthProvider and obtains a JWT.

2. Using the RoochSDK, the user generates a zero-knowledge proof that they possess a valid JWT without revealing the content of the JWT. This proof includes a proof and some public signals, such as `nonceValid`, `jwtValid`, and `roochAddress`.

3. The user creates a `CreateSessionKey` transaction using the RoochSDK, including the public key, zero-knowledge proof, public signals, expiration time, and the scopes that define the modules on the chain that the session key can access.

4. The RoochSDK sends the transaction to the RoochServer. The RoochServer uses the `JWTZKProofAuthValidator` validator to validate this transaction. If the transaction is invalid, the validator will throw an error.

5. If the transaction is valid, the RoochServer stores the public key and the associated scopes in the user's Rooch account. The user retains the private key for subsequent transactions.

6. The user can use their private key to initiate new transactions via the RoochSDK within the defined scopes until the Session Key expires.

This process ensures that users can authenticate their identity on the Rooch chain using their OAuthProvider account while protecting their privacy. Zero-knowledge proof allows users to prove that they have a valid JWT without revealing the content of the JWT. In addition, by using the Session Key and defining scopes, users can initiate multiple transactions over a period of time without having to authenticate their identity each time, and the access to the modules on the chain is controlled.

**Comment by yubing744** [comment]
The zero-knowledge proof circuit should include the following logic:

1. **JWT Verification**: The circuit needs to verify the validity of the JWT. This could include decoding the JWT, verifying the signature, checking the issuer and subject, etc.

2. **Nonce Verification**: The circuit needs to verify the validity of the nonce. This typically means checking if the nonce matches the expected value.

3. **Rooch Address Generation**: The circuit needs to generate a Rooch address based on the JWT. This could involve using certain parts of the JWT (such as the user ID) as input, and then generating the address through some hash or encoding algorithm.

4. **Public Signal Generation**: The circuit needs to generate some public signals, such as `nonceValid`, `jwtValid`, and the generated `roochAddress`. These signals can be verified externally to the circuit to ensure the correctness of the zero-knowledge proof.

**Comment by yubing744** [comment]
The zero-knowledge proof circuit specification:

The zero-knowledge circuit should accept the following input parameters:

1. JWT: The JSON Web Token obtained by the user from the OAuthProvider.

2. OAuthProvider Public Key: The public key used to verify the JWT signature.

3. Issuer Public Key: The public key used to verify the JWT issuer.

4. sequence_number: The sequence_number of the user's Rooch account, used to generate public signals and prevent replay attacks.

The zero-knowledge circuit should perform the following operations:

1. Verify the signature using the OAuthProvider's public key and the header and payload of the JWT.

2. Verify the issuer using the predefined issuer public key.

3. Verify whether the JWT has expired.

4. Generate `roochAddress` based on the information in the JWT.

5. Generate some public signals, such as `sequence_number`, `jwtValid`, and `roochAddress`.

The zero-knowledge circuit should output the following parameters:

1. Zero-knowledge proof: A proof that the user has a valid JWT, without revealing the content of the JWT.

2. Public signals: Including `sequence_number`, `jwtValid`, and `roochAddress`.

**Comment by yubing744** [comment]
Example Implements:

```
    include "../../node_modules/circomlib/circuits/eddsaproof.circom";
    include "../../node_modules/circomlib/circuits/comparators.circom";
    include "../../node_modules/circomlib/circuits/merklehash.circom";
    
    template VerifyJWT() {
      
      signal input jwt;
      signal input oauthPubKey;
      signal input issuerPubKey;
      signal input sequence_number;
      
      // 1. 验证签名
      component verifySignature = EDDsaVerify();
      verifySignature.pubKey <== oauthPubKey;
      verifySignature.signature <== jwt.signature; 
      verifySignature.msgHash <== jwt.header && jwt.payload;
      
      // 2. 验证签发者 
      component verifyIssuer = EdDSAPubKeyEq();
      verifyIssuer.in[0] <== issuerPubKey;
      verifyIssuer.in[1] <== jwt.header.iss;
      
      // 3. 验证是否过期
      component isExpired = LessThan(64);
      isExpired.in[0] <== jwt.payload.exp;
      isExpired.in[1] <== now;
      
      // 4. 生成roochAddress
      component generateAddress = Pedersen(21888242871839275222246405745257275088548364400416034343698204186575808495617);
      generateAddress.in <== sequence_number;
      
      // 5. 生成公共信号
      signal jwtValid = verifySignature.out && verifyIssuer.out && !isExpired.out;
      signal roochAddress = generateAddress.out;
      
      // 输出
      signal output zkProof;
      signal output sequence_number;
      signal output jwtValid;
      signal output roochAddress;
    }
    
    component main {public} = VerifyJWT();
```

**Comment by feliciss** [comment]
I feel like using Halo2 from zcash might be a more succinct solution in the future?

https://github.com/zcash/halo2

**Comment by yubing744** [comment]
> I feel like using Halo2 from zcash might be a more succinct solution in the future?
> 
> https://github.com/zcash/halo2

I considered snarkjs because the game Dark Forest also uses this library. snarkjs uses circom to write verification circuits and compile them into wasm files, which can generate proofs directly in the browser.

I don’t know much about Halo2, can you elaborate on the advantages of Halo2?

**Comment by feliciss** [comment]
> > I feel like using Halo2 from zcash might be a more succinct solution in the future?
> > https://github.com/zcash/halo2
> 
> I considered snarkjs because the game Dark Forest also uses this library. snarkjs uses circom to write verification circuits and compile them into wasm files, which can generate proofs directly in the browser.
> 
> I don’t know much about Halo2, can you elaborate on the advantages of Halo2?

Halo2 is a PLONKish arithmetization scheme zero-knowledge proof which produces larger size of proofs than Groth16, does not require trusted setups and can be used to customize custom gates and lookup tables.

There's an article describing the efficiency of each ZKP frameworks, PLONKish, R1CS, etc.:

https://ethresear.ch/t/benchmarking-zkp-development-frameworks-the-pantheon-of-zkp/14943

Here's an implementation using WASM with Halo2:

https://github.com/axiom-crypto/halo2-wasm/tree/main

**Comment by feliciss** [comment]
I may help design with you together for this issue. @yubing744

**Comment by yubing744** [comment]
@feliciss Does Sui use Halo2 or circom? I found that zkemail(https://github.com/zkemail/) also tried to adopt Halo2

**Comment by yubing744** [comment]
@jolestar sui supports [Zero-knowledge proof verification (Groth16)](https://github.com/MystenLabs/sui/blob/c25689f5c142b495c05306d30fac847bb53ca244/doc/src/learn/cryptography/groth16.md?plain=1#L27)

**Comment by yubing744** [comment]
RSA verification circuit implementation reference:
* https://github.com/zkp-application/circom-rsa-verify
* https://github.com/zkemail/zk-email-verify/blob/main/packages/circuits/helpers/rsa.circom
* https://github.com/dmpierre/zkrsa

**Comment by yubing744** [comment]
@jolestar @feliciss I have completed the JWT validation circuit and passed the unit test.

https://github.com/yubing744/rooch/blob/feature-owen-zklogin-circuit-verify/sdk/zklogin/circuits/zklogin/tests/jwt.test.ts

**Comment by feliciss** [comment]
> @feliciss Does Sui use Halo2 or circom? I found that zkemail(https://github.com/zkemail/) also tried to adopt Halo2

No. Sui uses Groth16 that's different from PLONKish implementation. Sorry for the late reply.

**Comment by yubing744** [comment]
> > @feliciss Does Sui use Halo2 or circom? I found that zkemail(https://github.com/zkemail/) also tried to adopt Halo2
> 
> No. Sui uses Groth16 that's different from PLONKish implementation. Sorry for the late reply.

circom also supports generating Groth16 proofs. We plan to implement Rooch’s Groth16 verification with reference to Sui.
